lib/dialect/src/ConstraintsAnalysis.cpp

```diff
@@ -0,0 +1,1028 @@
+
+				// We can only do this analysis if the function return true or false
+				if(auto bool_branch=mlir::dyn_cast<BoolAttr>(value.getDefiningOp()->getAttr("value"))){
+					// Insert in the lattice the constant I have found
+					// We can insert an empty range since this is used just for check
+					this->ranges.insert(std::make_pair(
+						std::make_pair(value,bool_branch.getValue()),
+						this->getBottom() ));
+					
+					return mlir::ChangeResult::Change;
+				}
+
+				return mlir::ChangeResult::NoChange;
+			}
+
+			// TODO: divide this operation in (at least) two parts
+			// Here we should write the transfer function based on the operation 
+			mlir::ChangeResult visitArithmeticOperation(mlir::Operation* statement, bool can_true, bool can_false)
+			{
+
+				// Modify my ranges only when i am sure that i am passing through a initialized (true/false) branch
+				if (not can_true and not can_false)
+					return mlir::ChangeResult::NoChange;
+
+				// Here the operation has passed the sanity check so we can switch
+				auto copy = *this;
+
+				// So this becomes extremely tricky now, we have the information about the branch we are passing from
+				// The idea is, i am visiting one branch => update one, I am visiting both => update both
+
+				// TODO: this should be a casting to a CommutativeOp <- TODO: add
+				// For the moment it should work so let's leave it like this
+				if (auto casted=mlir::dyn_cast<mlir::rlc::AddOp>(statement)){
+					llvm::outs()<<"addop\n";
+					this->print(llvm::outs());
+					// Do some name changing so that hopefully is more clear
+					// NB : these are values
+					const auto& a = casted.getResult();
+					const auto& b = casted.getLhs();
+					const auto& c = casted.getRhs();
+					// This can be nullptr if for example it is a function argument
+					// TODO: are there any other cases ?
+					const auto& b_op = b.getDefiningOp();
+					const auto& c_op = c.getDefiningOp();
+					// TODO: understand what happens if both have ranges -> should I update both ? (because if yes, in which order ?)
+					// NB: there iConstraintsLattices the case in which both are non constantv-> if we know nothing about
+					//     both we do nothing
+					// Then first find which variable we need to operate on
+					mlir::Value      state=nullptr;
+					mlir::Value      stato=nullptr;
+					// First get our candidates
+					if(b_op == nullptr or not mlir::isa<mlir::rlc::Constant>(b_op)){
+						state=b;
+						stato=c;
+					}
+					else if(c_op == nullptr or not mlir::isa<mlir::rlc::Constant>(c_op)){
+						state=c;
+						stato=b;
+					}
+					// Then update the corresponding branch
+					// NB: they are not exclusive so I have to update both
+					if(can_true)
+						this->updateArithmeticCommutativeRange(statement, a, state, stato, true);
+					if(can_false)
+						this->updateArithmeticCommutativeRange(statement, a, state, stato, false);
+
+				}
+
+				if(*this!=copy){
```

Comment:
uhm, we can't copy and check if there has been no change at every operation, this would get expensive pretty fast, i guess insertOrJoin should return if the inserted range modified the original content instead?

Reply:
hm probably its better yes, it will need some refactoring since we will have to carry on that information for a lot of the function lifetime. But definitely better a boolean than an entire object